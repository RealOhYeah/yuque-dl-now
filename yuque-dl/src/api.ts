import { env } from 'node:process'
import axios from 'axios'
import { randUserAgent } from './utils'
import { DEFAULT_COOKIE_KEY, DEFAULT_DOMAIN } from './constant'

import type {
  ArticleResponse,
  KnowledgeBase,
  GetHeaderParams,
  IReqHeader,
  TGetKnowledgeBaseInfo,
  TGetMdData
} from './types'
import type { AxiosRequestConfig } from 'axios'

function getHeaders(params: GetHeaderParams): IReqHeader {
  const { key = DEFAULT_COOKIE_KEY, token } = params
  const headers: IReqHeader = {
    'user-agent': randUserAgent({
      browser: 'chrome',
      device: 'desktop'
    })
  }
  if (token) headers.cookie = `${key}=${token};`
  return headers
}

export function genCommonOptions(params: GetHeaderParams): AxiosRequestConfig {
  const config: AxiosRequestConfig = {
    headers: getHeaders(params),
    beforeRedirect: (options) => {
      // 语雀免费非企业空间会重定向如: www.yuque.com -> gxr404.yuque.com
      // 此时axios自动重定向并不会带上cookie
      options.headers = {
        ...(options?.headers || {}),
        ...getHeaders(params)
      }
    }
  }
  if (env.NODE_ENV === 'test') {
    config.proxy = false
  }
  return config
}


/** 获取知识库数据信息 */
export const getKnowledgeBaseInfo: TGetKnowledgeBaseInfo = (url, headerParams) => {
  const knowledgeBaseReg = /decodeURIComponent\("(.+)"\)\);/m
  return axios.get<string>(url, genCommonOptions(headerParams))
    .then(({data = '', status}) => {
      if (status === 200) return data
      return ''
    })
    .then(html => {
      const data = knowledgeBaseReg.exec(html) ?? ''
      if (!data[1]) return {}
      const jsonData: KnowledgeBase.Response = JSON.parse(decodeURIComponent(data[1]))
      if (!jsonData.book) return {}
      const info = {
        bookId: jsonData.book.id,
        bookSlug: jsonData.book.slug,
        tocList: jsonData.book.toc || [],
        bookName: jsonData.book.name || '',
        bookDesc: jsonData.book.description || '',
        host: jsonData.space?.host || DEFAULT_DOMAIN,
        imageServiceDomains: jsonData.imageServiceDomains || []
      }
      return info
    }).catch((e) => {
      // console.log(e.message)
      const errMsg = e?.message ?? ''
      if (!errMsg) throw new Error('unknown error')
      const netErrInfoList = [
        'getaddrinfo ENOTFOUND',
        'read ECONNRESET',
        'Client network socket disconnected before secure TLS connection was established'
      ]
      const isNetError = netErrInfoList.some(netErrMsg => errMsg.startsWith(netErrMsg))
      if (isNetError) {
        throw new Error('请检查网络(是否正常联网/是否开启了代理软件)')
      }
      throw new Error(errMsg)
    })
}


export const getDocsMdData: TGetMdData = (params, isMd = true) => {
  const { articleUrl, bookId, token, key, host = DEFAULT_DOMAIN } = params
  let apiUrl = `${host}/api/docs/${articleUrl}`
  const queryParams: any = {
    'book_id': String(bookId),
    'merge_dynamic_data': String(false)
    // plain=false
    // linebreak=true
    // anchor=true
  }
  if (isMd) queryParams.mode = 'markdown'
  const query = new URLSearchParams(queryParams).toString()
  apiUrl = `${apiUrl}?${query}`
  return axios.get<ArticleResponse.RootObject>(apiUrl, genCommonOptions({token, key}))
    .then(({data, status}) => {
      const res = {
        apiUrl,
        httpStatus: status,
        response: data
      }
      return res
    })
}
