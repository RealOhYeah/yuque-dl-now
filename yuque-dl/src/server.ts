import { createServer } from 'vitepress'
import { mkdir, writeFile, readFile, readdir, stat, access } from 'node:fs/promises'
import { join, resolve, sep } from 'node:path'

let restartPromise: Promise<void> | undefined

export async function runServer(root: string) {
  const rootPath = resolve(root)
  if (!await fileExists(rootPath)) {
    throw new Error('server root not found')
  }
  await createVitePressConfig(rootPath)
  const createDevServer = async () => {
    const server = await createServer(root, {}, async () => {
      if (!restartPromise) {
        restartPromise = (async () => {
          await server.close()
          await createDevServer()
        })().finally(() => {
          restartPromise = undefined
        })
      }

      return restartPromise
    })
    await server.listen()
    server.printUrls()
    // bindShortcuts(server, createDevServer)
  }

  await createDevServer().catch((err) => {
    console.error(err)
    process.exit(1)
  })
}

async function createVitePressConfig(root: string) {
  const bookName = root.split(sep).filter(Boolean).at(-1) || 'yuque-dl'
  const vitepressPath = join(root, '/.vitepress')
  await mkdir(vitepressPath, {recursive: true})
  const vitePressConfig = join(vitepressPath, 'config.ts')
  const sidebar = await createSidebarMulti(root)
  const config = `
  export default {
    title: "${bookName}",
    themeConfig: {
      search: {
        provider: 'local'
      },
      sidebar: ${JSON.stringify(sidebar)}
    },
    vite: {
      optimizeDeps: {
        include: []
      }
    },
    markdown: {
      html: false
    }
  }
  `
  await writeFile(vitePressConfig, config)
}

async function fileExists(filename: string) {
  try {
    await access(filename)
    return true
  } catch (err) {
    if (err.code === 'ENOENT') {
      return false
    } else {
      throw err
    }
  }
}

async function createSidebarMulti (path: string) {
  const data = [] as any
  const ignoreList = ['.vitepress', 'img', 'index.md', 'progress.json', 'attachments']
  let dirList = await readdir(path)
  dirList = dirList.filter(item => !ignoreList.includes(item))
  for (const n of dirList) {
    const dirPath = join(path, n)
    const statRes = await stat(dirPath)
    if (statRes.isDirectory()) {
      const isHasIndex = await fileExists(join(dirPath,'index.md'))
      const item: any = {
        text: n,
        collapsed: true,
        items: await createSideBarItems(path, n)
      }
      if (isHasIndex) item.link = `/${n}/`
      data.push(item)
    } else {
      data.push({
        text: n.slice(0, n.lastIndexOf('.')),
        link: `/${n}`
      })
    }
  }
  return data
}

// 尝试从一个md文件中读取标题，读取到第一个 ‘# 标题内容’ 的时候返回这一行
export async function getTitleFromFile (realFileName: string): Promise<string | undefined> {
  const isExist = await fileExists(realFileName)
  if (!isExist) {
    return undefined
  }
  const fileExtension = realFileName.substring(
    realFileName.lastIndexOf('.') + 1
  )
  if (fileExtension !== 'md' && fileExtension !== 'MD') {
    return undefined
  }
  // read contents of the file
  const data = await readFile(realFileName, { encoding: 'utf-8' })
  // split the contents by new line
  const lines = data.split(/\r?\n/)
  // return title
  for (const line of lines) {
    if (line.startsWith('# ')) {
      return line.substring(2)
    }
  }
  return undefined
}

async function createSideBarItems (
  targetPath: string,
  ...reset: string[]
) {
  const collapsed = false
  const node = await readdir(join(targetPath, ...reset))

  const result = []
  for (const fname of node) {
    const curPath = join(targetPath, ...reset, fname)
    const statRes = await stat(curPath)
    if (statRes.isDirectory()) {
      const items = await createSideBarItems(join(targetPath), ...reset, fname)
      if (items.length > 0) {
        const sidebarItem: any = {
          text: fname,
          items
        }
        // vitePress sidebar option collapsed
        sidebarItem.collapsed = collapsed
        const isHasIndex = await fileExists(join(curPath, 'index.md'))
        if (isHasIndex) {
          sidebarItem.link = '/' + [...reset, fname].map(decodeURIComponent).join('/') + '/'
        }
        result.push(sidebarItem)
      }
    } else {
      // is filed
      if (
        fname === 'index.md' ||
        /^-.*\.(md|MD)$/.test(fname) ||
        !fname.endsWith('.md')
      ) {
        continue
      }
      const fileName = fname.replace(/\.md$/, '')
      const item = {
        text: fileName,
        link: '/' + [...reset.map(decodeURIComponent), `${encodeURI(fileName)}.html`].join('/')
      }
      result.push(item)
    }
  }
  return result
}
