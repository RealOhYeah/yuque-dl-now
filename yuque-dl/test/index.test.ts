import path from 'node:path'
import { afterAll, afterEach, beforeAll, beforeEach, describe, expect, it } from 'vitest'
import { TestTools } from './helpers/TestTools'
import { server } from './mocks/server'
import { KNOWLEDGE_BASE_URL } from './helpers/constant'
import { main } from '../src/index'
import { readdirSync, readFileSync } from 'node:fs'


let testTools: TestTools

describe('main', () => {
  beforeAll(() => {
    server.listen({ onUnhandledRequest: 'error' })
  })
  afterAll(() => server.close())

  beforeEach(() => {
    testTools = new TestTools()
  })

  afterEach(() => {
    testTools.cleanup()
    server.resetHandlers()
  })
  it('should work', async () => {
    await main(KNOWLEDGE_BASE_URL.BASE1, {
      distDir: testTools.cwd,
      ignoreImg: false,
      toc: false
    })
    const summaryPath = path.join(testTools.cwd, '知识库TEST1/index.md')
    expect(readFileSync(summaryPath).toString()).toMatchSnapshot()
    const doc2 = path.join(testTools.cwd, '知识库TEST1/Title2/文档2.md')
    expect(readFileSync(doc2).toString()).toMatchSnapshot()
    const progressJSON = path.join(testTools.cwd, '知识库TEST1/progress.json')
    expect(readFileSync(progressJSON).toString()).toMatchSnapshot()
    let doc1Data = readFileSync(path.join(testTools.cwd, '知识库TEST1/Title1/文档1.md')).toString()
    doc1Data = doc1Data.replace(/\.\/img.*?-(\d{6})\./g, (match, random) => {
      return match.replace(random, '123456')
    })
    expect(doc1Data).toMatchSnapshot()
    const imgDir = path.join(testTools.cwd, '知识库TEST1/Title1/img/002')
    const imgList = readdirSync(imgDir)
    const img1 = path.join(testTools.cwd, `知识库TEST1/Title1/img/002/${imgList[0]}`)
    expect(readFileSync(img1).length).toBe(99892)
    const img2 = path.join(testTools.cwd, `知识库TEST1/Title1/img/002/${imgList[1]}`)
    expect(readFileSync(img2).length).toBe(81011)
  })
})
