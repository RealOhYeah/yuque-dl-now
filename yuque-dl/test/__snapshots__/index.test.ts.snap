// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`main > should work 1`] = `
"# 知识库TEST1

> 知识库 test desc


## Title1

- [文档1](Title1/文档1.md)

## Title2

- [文档2](Title2/文档2.md)
"
`;

exports[`main > should work 2`] = `
"# 文档2

# DOC2
## Title
text...
### Title2
text...


> 原文: <https://www.yuque.com/yuque/base1/two>"
`;

exports[`main > should work 3`] = `"[{"path":"Title1","pathTitleList":["Title1"],"pathIdList":["001"],"toc":{"type":"TITLE","title":"Title1","uuid":"001","child_uuid":"002","parent_uuid":""}},{"path":"Title1/文档1.md","savePath":"Title1","pathTitleList":["Title1","文档1"],"pathIdList":["001","002"],"toc":{"type":"DOC","title":"文档1","uuid":"002","url":"one","child_uuid":"","parent_uuid":"001"}},{"path":"Title2","pathTitleList":["Title2"],"pathIdList":["003"],"toc":{"type":"TITLE","title":"Title2","uuid":"003","child_uuid":"004","parent_uuid":""}},{"path":"Title2/文档2.md","savePath":"Title2","pathTitleList":["Title2","文档2"],"pathIdList":["003","004"],"toc":{"type":"DOC","title":"文档2","uuid":"004","url":"two","child_uuid":"","parent_uuid":"003"}}]"`;

exports[`main > should work 4`] = `
"# 文档1

# DOC1
![1.jpeg](./img/002/1-123456.jpeg)
## SubTitle
![2.jpeg](./img/002/2-123456.jpeg)

> 原文: <https://www.yuque.com/yuque/base1/one>"
`;
